### ISSUE TITLE
probability of result 

### ISSUE THREAD

**Issue by mohammad-karam** [haverland/Tenth-of-step-of-a-meter-digit#5]
Hi @haverland,

The output of the NN is an 2d array with shape [100,2] the answer is the index with largest activation. It is usually between -1 and 15 . How could I convert activation number to percentage?  to be able to say prediction e.g. is 98(9.8) with a probability 90%.

**Comment by haverland** [comment]
I don't know which notebook you mean with NN. All notebooks (models) has an output of 100 here. So it uses classical categorical np.argmax() / 10, because categorical needs integer for encoding.

I've used an encoding function, because I played with different model types. (b2n/encodings/class_encoding.py). But It is as the same as keras.utils.to_categorical.

So look on you output shape. I think it should be [1, 100]. 

If you look at b2n/plotfunctions.py, where is an example of prediction of normal model (predict_meter_digits) and quantized (tflite) model is the function evaluate_ziffer_tflite.

Hope it helps.

**Comment by haverland** [comment]
If you only looking for an implementation for https://github.com/jomjol/AI-on-the-edge-device look at the rolling branch. It can use the tflite-model from here

**Comment by mohammad-karam** [comment]
Thanks for your quick answer.

sorry for my mistake it is 1d array [100,1] . I have looked to the b2n/plotfunctions.py you were calculating the accuracy by dividing the false predicted photos over all photos. 

In my Scenario  I have only one photo and I want to predict it and I want to get the accuracy of prediction . I have used your trained model output/cnn-s1-md_best.h5 using the argmax ()/10 I got the answer . When I looked to the activation value of  argmax () it is usually between (1 -  13) . in other models this number (1- 13 ) represents the  accuracy of prediction .  I mean when we have only two classes for example (Orange, apple ) you get usually output with shape [2,1] with values [80 ,20 ], where first element represent the percentage of accuracy and so one  . 
The result I want to show is (Orange 70 percent).

I searched in Internet to convert the activations of last layer to percentage, AFAIK you may need to set an activation method for your last layer.

https://stackoverflow.com/questions/1523420/how-to-convert-the-output-of-an-artificial-neural-network-into-probabilities

**Comment by haverland** [comment]
In output, you get every time an array [-1, 100, 1] or [-1, 100]. As input you have to use the same. With one image is the shape [1, 32, 20, 3]. That's why the output is also an array [1, 100, 1]. 

I've used tf.keras.losses.CategoricalCrossentropy(from_logits=True) instead of softmax. With np.argmax the result is the same. Softmax makes a percentage between 0..1 over all output. Means the sum of all outputs is 1. 

import numpy as np
def softmax(x):
    """Compute softmax values for each sets of scores in x."""
    e_x = np.exp(x - np.max(x))
    return e_x / e_x.sum(axis=0) # only difference
    
yourimage = loadmyimage()...
yourimage = np.expand_dims(yourimage)
classes = model.predict(yourimage)
predictions = np.argmax(classes, axis=1).reshape(-1)/10

softmax_pred = softmax(predictions) # the sum(soft_pred) on one image = 1

**Comment by mohammad-karam** [comment]
If I use the `softmax  ` method and consider the difference as a sign of reliability, I will have a big problem with a non digit images ,Because such images can return a big difference between the outputs and as result a high reliability with a wrong result. 

But It is good way if I were sure that the image is a digit.

**Comment by haverland** [comment]
This project have no "non digit" classification like jomols models. It returns every time a digit result. Big differences and negatives will be handled in post-process of the [AI-on-the-edge-device](https://github.com/jomjol/AI-on-the-edge-device).

Means the model must be have a quite good accuracy and if it false predict, the post process will run in an error (rate to high or negative). This works very well for me on 4 meters.

**Comment by haverland** [comment]
closed

**Comment by haverland** [comment]
closed
